src/coreclr/vm/method.cpp

```diff
@@ -1205,8 +1205,15 @@ COR_ILMETHOD* MethodDesc::GetILHeader()
 
     Module *pModule = GetModule();
 
+    TADDR pIL = 0;
     // Always pickup overrides like reflection emit, EnC, etc.
-    TADDR pIL = pModule->GetDynamicIL(GetMemberDef());
+    // Unless it is an async thunk.
+    // A thunk cannot be overriden, bur shares methoddef with the real method,
+    // which could be overriden, but we do not want that override.
+    if (!IsAsyncThunkMethod())
```

Comment:
It does not look like right that we check IsAsyncThunkMethod here to deal with reflection.emit case and in GetRVA to deal with non-reflection.emit case. Can the check from `GetRVA()` be moved up so that it is in just one place? 

I see checked for `IsUnboxingStub` in `BOOL HasILHeader()`. Should this mirror checks for unboxing stubs? Unboxing stubs has same problem - the real method IL should not be used for them.

Reply:
I have noticed that `MethodDescGetRVA` method is unused. You can delete it while you are on it.

Reply:
We have moved these checks between `GetRVA` and `HasILHeader` once already. I think the reason is that result of `HasILHeader` depends on whether a method has a body. `GetRVA` is a lower method, but is not always called - i.e. is not called for dynamic overrides. 
Thus neither `GetRVA` nor `HasILHeader` is a good single place for these checks.

I think it can be made more clear if we add `MayHaveILHeader` that only tells if a method may possibly have a header. 
Non-IL methods and methods whose implementation is a transient piece of IL (i.e. unbox stubs and async thunks) would return `false`. 
Such method could be a single place for the checks.

Reply:
Yes, the semantics of these methods are a mess. It caused problems number of times...